arangod/Pregel/Actor/include/Actor/MPSCQueue.h

```diff
@@ -0,0 +1,130 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <memory>
+#include <atomic>
+
+// This lock-free multi-producer-single-consumer-queue is inspired by the
+// code published at
+//
+// https://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue
+//
+// under a Apache-2.0 license.
+namespace arangodb::pregel::mpscqueue {
+
+template<typename T>
```

Comment:
You need to remove this code and use "queue" which allow you to call kick() not on every process.

Or modify actor code, to not call kick on every proccess with this queue.
But in such case you also should use weaker memory orders here

Reply:
Thanks, first part done.

Now please do second part, specify memory_orders or remove ineffective lock-free code.
If you don't know how, do you think is good idea to merge such code in arangodb?

Reply:
It can be true that the code is more efficient if we specify the memory orders. But it is a lot of work to figure out the most efficient and correct memory order for each atomic operation and we don't even have a performance test that shows that this effort would be worth the time spent. Therefore I would like to stick to the sequentially consistent memory order for now which will give correct results.